Fixes #22789 - Add structured data as a response

xprazak2 opened the pull request.

ares commented:
this no needs rebasing, hopefully it won't be too bad

xprazak2 commented:
Rebased, still need to restructure the response into a separate object per template instead of list on export.

ares commented:
Also this seems to break rubocop

```
RuboCop failed!
Running RuboCop...
Inspecting 16 files
......C.........

Offenses:

/var/lib/workspace/workspace/test_plugin_pull_request/database/sqlite3/label/fast/ruby/2.4/plugin/app/services/foreman_templates/parse_result.rb:59:23: C: Naming/UncommunicativeMethodParamName: Method parameter must be at least 3 characters long.
    def add_exception(ex)
                      ^^
/var/lib/workspace/workspace/test_plugin_pull_request/database/sqlite3/label/fast/ruby/2.4/plugin/app/services/foreman_templates/parse_result.rb:65:20: C: Naming/UncommunicativeMethodParamName: Method parameter must be at least 3 characters long.
    def name_error(ex)
                   ^^

```

xprazak2 commented:
Rebased, showing templates as objects instead of just list of names.

xprazak2 commented:
I changed the message for template that is filtered out.

xprazak2 commented:
Updated, I changed the output structure for export. 'finish' and 'PXELinux' are template kinds, these templates are also provisioning templates, so I think the structure you suggest might be a bit confusing.

Now we have:

```js
{ 'templates':
     'provisioning_template': [ 'template_a'. 'template_b']
     'ptable': ['ptable_a', 'ptable_b']
     'job_template': ['job_template_a', 'job_template_b']
}
```

I also tried to reproduce the issue with `diff` and `changed` not showing up properly for `verbose`, but it works for me:

![](https://i.imgur.com/rSd2rMg.png)

xprazak2 commented:
I managed to fix the number 4, here as well.

xprazak2 commented on app/services/foreman_templates/action.rb:
```diff
@@ -22,17 +22,6 @@ def initialize(args = {})
       assign_attributes args
     end
 
-    def get_default_branch(repo)
```
It is not related to the response format, but the method relies on the naming that we have in our repositories. I removed it because I already make changes in `TemplateExporter#export!` and I see no value in keeping it.

xprazak2 commented:
Now we have for import:

```
{
    "message": {
        "templates": [
            {
                "name": "A fake pxelinux",
                "id": 202,
                "changed": false,
                "imported": true,
                "additional_errors": null,
                "exception": null,
                "validation_errors": {

                },
                "file": "a_fake_pxelinux.erb",
                "type": "provisioning_template"
            },
            {
                "name": "A fake provision",
                "id": 203,
                "changed": false,
                "imported": true,
                "additional_errors": null,
                "exception": null,
                "validation_errors": {

                },
                "file": "a_fake_provision.erb",
                "type": "provisioning_template"
            },
...
```

and export:

```
{
    "message": {
        "error": null,
        "warning": null,
        "repo": "ssh://git@github.com/xprazak2/fake_templates.git",
        "branch": "whktml",
        "git_user": "admin",
        "templates": {
            "provisioning_template": [
                {
                    "id": 9,
                    "name": "PXELinux chain iPXE UNDI",
                    "exported": true
                },
                {
                    "id": 99,
                    "name": "Katello Atomic Kickstart Default",
                    "exported": true
                },
...
```

I can't very well do grouping by template class on import, because some templates may be missing model in metadata and having them in group `nil_class` is not very nice. That leaves us with slightly different structure or we can just have all the templates in one array as on import.

I also moved some logic aggregating the export results into `ExportResult` class.

ares commented:
Ok, now I see the import limitation. I think it would be worth of having export not categorized with the template type then, I think the API structure should be as consistent as possible. Since you now added more top level keys to export, can they also be added to import? @jyejare any concern with this change?

xprazak2 commented:
Updated, I added repo and branch to the top level output for the import and removed the grouping by type for export.

ares commented on app/services/foreman_templates/template_importer.rb:
```diff
@@ -33,94 +34,97 @@ def import_from_git
         gitrepo = Git.clone(@repo, @dir)
-        branch = @branch ? @branch : get_default_branch(gitrepo)
-        gitrepo.checkout(branch) if branch
+        gitrepo.checkout(@branch) if @branch
```
If @branch is loaded from settings and it's empty string, this will fail I'm afraid. Since we dropped get_default_branch we should either add `@branch.blank?` or fallback to "master"

and I think we should reuse `TemplateExporter#setup_git_branch` so perhaps move it to action and reuse here?

ares commented:
also this now needs rebase

xprazak2 commented:
Rebased and fixing tests.

ares commented:
@xprazak2 it seems the tests are still failing - http://ci.theforeman.org/job/foreman_templates-pull-request/2/database=mysql,label=fast,ruby=2.4/testReport/junit/(root)/ForemanTemplates__TemplateImporterTest__other%20plugins/test_0001_are_ignored_if_not_installed/

ares commented:
Thanks @xprazak2, works great and API output seems as consistent as possible.
